### PR TITLE
Move vsphere machine config validations to api

### PR DESCRIPTION
**Pull request by taneyland**
*Issue #, if available:*

*Description of changes:*
This PR moves some vsphere machine config validations to the api level. This is especially important for the full lifecycle api to validate the fields during both create and upgrade through the controller

*Testing (if applicable):*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: pkg/api/v1alpha1/vspheremachineconfig.go [hunk 1]
```diff
@@ -76,3 +78,23 @@ func GetVSphereMachineConfigs(fileName string) (map[string]*VSphereMachineConfig
 	}
 	return configs, nil
 }
+
+func validateVSphereMachineConfig(config *VSphereMachineConfig) error {
+	if len(config.Spec.Datastore) <= 0 {
+		return fmt.Errorf("VSphereMachineConfig %s datastore is not set or is empty", config.Name)
+	}
+	if len(config.Spec.Folder) <= 0 {
+		logger.Info("VSphereMachineConfig folder is not set or is empty. Will default to root vSphere folder.")
```
Comment (g-gaston) [inline pkg/api/v1alpha1/vspheremachineconfig.go:87]: I'm not sure this log belongs here, could we find a better place? For starters this is talking about some defaulting logic, while this method runs validations

File: pkg/providers/vsphere/validator.go [hunk 2]
```diff
@@ -107,17 +107,8 @@ func (v *Validator) ValidateClusterMachineConfigs(ctx context.Context, vsphereCl
...
-	}
-	if len(controlPlaneMachineConfig.Spec.Folder) <= 0 {
-		logger.Info("VSphereMachineConfig folder for control plane is not set or is empty. Will default to root vSphere folder.")
-	}
-	if len(controlPlaneMachineConfig.Spec.ResourcePool) <= 0 {
-		return errors.New("VSphereMachineConfig VM resourcePool for control plane is not set or is empty")
-	}
-	if controlPlaneMachineConfig.Spec.OSFamily != anywherev1.Bottlerocket && controlPlaneMachineConfig.Spec.OSFamily != anywherev1.Ubuntu {
-		return fmt.Errorf("control plane osFamily: %s is not supported, please use one of the following: %s, %s", controlPlaneMachineConfig.Spec.OSFamily, anywherev1.Bottlerocket, anywherev1.Ubuntu)
+	if err := controlPlaneMachineConfig.Validate(); err != nil {
```
Comment (g-gaston) [inline pkg/providers/vsphere/validator.go:110]: Do we need to run this here? Shouldn't this already be run when we validate the `cluster.Config`?

File: pkg/providers/vsphere/validator.go [hunk 3]
```diff
@@ -694,3 +635,31 @@ func (v *Validator) getMissingPrivs(ctx context.Context, vsc govmomi.VSphereClie
 
 	return missingPrivs, nil
 }
+
+func (v *Validator) sameOSFamily(configs map[string]*anywherev1.VSphereMachineConfig) bool {
+	var osFamily anywherev1.OSFamily
+	for _, machineConfig := range configs {
+		if osFamily == "" {
+			osFamily = machineConfig.Spec.OSFamily
```
Comment (g-gaston) [inline pkg/providers/vsphere/validator.go:643]: Could we do this before the loop by assigning it to the first value?

File: pkg/providers/vsphere/validator.go [hunk 4]
```diff
@@ -694,3 +635,31 @@ func (v *Validator) getMissingPrivs(ctx context.Context, vsc govmomi.VSphereClie
...
+		}
+	}
+	return true
+}
+
+func (v *Validator) sameTemplate(configs map[string]*anywherev1.VSphereMachineConfig) bool {
+	var template string
+	for _, machineConfig := range configs {
+		if template == "" {
+			template = machineConfig.Spec.Template
```
Comment (g-gaston) [inline pkg/providers/vsphere/validator.go:657]: Same comment here about doing this out the loop

**Review by g-gaston — COMMENTED** [review]
Looks good!

File: pkg/api/v1alpha1/vspheremachineconfig.go (hunk 1 above)
Comment (taneyland) [reply pkg/api/v1alpha1/vspheremachineconfig.go]: I'll move it to the method where we set defaults. That will be in a separate PR coming soon

File: pkg/providers/vsphere/validator.go (hunk 2 above)
Comment (taneyland) [reply pkg/providers/vsphere/validator.go]: True, we don't need it here. Removing

File: pkg/providers/vsphere/validator.go (hunk 3 above)
Comment (taneyland) [reply pkg/providers/vsphere/validator.go:626]: I thought the same thing, but the object that we're traversing through is a map so I don't think there's a good way to access only the first object since we don't know what the first key will be. We could pass in the `OSFamily` of the control plane machine config to use as a source of truth for this, but I think this method would be nice to reuse in our `VSphereReconciler` when we validate the machine configs so I think this is the best way to satisfy both cases

File: pkg/providers/vsphere/validator.go [hunk 5]
```diff
@@ -188,51 +158,12 @@ func (v *Validator) ValidateClusterMachineConfigs(ctx context.Context, vsphereCl
...
-				return fmt.Errorf("validating SSHUsername for etcd VSphereMachineConfig %v: %v", etcdMachineConfig.Name, err)
-			}
-		}
-	} else {
-		return fmt.Errorf("validating SSHUsername for control plane VSphereMachineConfig %v: %v", controlPlaneMachineConfig.Name, err)
-	}
-
-	for _, machineConfig := range vsphereClusterSpec.machineConfigsLookup {
-		if machineConfig.Namespace != vsphereClusterSpec.Cluster.Namespace {
-			return errors.New("VSphereMachineConfig and Cluster objects must have the same namespace specified")
```
Comment (taneyland) [inline pkg/providers/vsphere/validator.go:216]: Removing because this is validated [here](https://github.com/aws/eks-anywhere/blob/main/pkg/cluster/vsphere.go#L56)

File: pkg/providers/vsphere/validator.go [hunk 6]
```diff
@@ -107,25 +107,9 @@ func (v *Validator) ValidateClusterMachineConfigs(ctx context.Context, vsphereCl
...
-		return errors.New("VSphereMachineConfig VM resourcePool for control plane is not set or is empty")
-	}
-	if controlPlaneMachineConfig.Spec.OSFamily != anywherev1.Bottlerocket && controlPlaneMachineConfig.Spec.OSFamily != anywherev1.Ubuntu {
-		return fmt.Errorf("control plane osFamily: %s is not supported, please use one of the following: %s, %s", controlPlaneMachineConfig.Spec.OSFamily, anywherev1.Bottlerocket, anywherev1.Ubuntu)
-	}
-
-	workerNodeGroupConfigs := vsphereClusterSpec.Cluster.Spec.WorkerNodeGroupConfigurations
-	for _, workerNodeGroupConfig := range workerNodeGroupConfigs {
-		if workerNodeGroupConfig.Name == "" {
-			return errors.New("must specify name for worker nodes")
```
Comment (taneyland) [inline pkg/providers/vsphere/validator.go:126]: Removing because it is already validated [here](https://github.com/aws/eks-anywhere/blob/main/pkg/api/v1alpha1/cluster.go#L409)

**Review by mitalipaygude — APPROVED** [review]

File: pkg/providers/vsphere/validator.go [hunk 7]
```diff
@@ -179,51 +148,12 @@ func (v *Validator) ValidateClusterMachineConfigs(ctx context.Context, vsphereCl
...
 
-	if etcdMachineConfig != nil && etcdMachineConfig.Spec.OSFamily != anywherev1.Bottlerocket && etcdMachineConfig.Spec.OSFamily != anywherev1.Ubuntu {
-		return fmt.Errorf("etcd node osFamily: %s is not supported, please use one of the following: %s, %s", etcdMachineConfig.Spec.OSFamily, anywherev1.Bottlerocket, anywherev1.Ubuntu)
-	}
-
-	if etcdMachineConfig != nil && controlPlaneMachineConfig.Spec.OSFamily != etcdMachineConfig.Spec.OSFamily {
-		return errors.New("control plane and etcd machines must have the same osFamily specified")
-	}
-
-	if err := v.validateSSHUsername(controlPlaneMachineConfig); err == nil {
```
Comment (mitalipaygude) [inline pkg/providers/vsphere/validator.go:190]: Were these validations moved too in a different PR?

File: pkg/providers/vsphere/validator.go (hunk 5 above)
Comment (mitalipaygude) [reply pkg/providers/vsphere/validator.go:207]: is this link the right one?

Comment (taneyland) [reply pkg/providers/vsphere/validator.go:207]: Right file, wrong line number. Updated it!

File: pkg/providers/vsphere/validator.go (hunk 7 above)
Comment (taneyland) [reply pkg/providers/vsphere/validator.go:190]: No, I moved it [here](https://github.com/aws/eks-anywhere/blob/9192960b13e7a0fe95f40abfb2e43cac89e023be/pkg/api/v1alpha1/vspheremachineconfig.go#L137) in this same PR. I just moved it out of the method since it is just one validation that can stand alone like the others

File: pkg/providers/vsphere/validator.go (hunk 3 above)
Comment (g-gaston) [reply pkg/providers/vsphere/validator.go:616]:
```go
for _, machineConfig := range configs {
    osFamily = machineConfig.Spec.OSFamily
    break
}
```

Comment (taneyland) [reply pkg/providers/vsphere/validator.go]: Yeah I considered doing this before. I personally don't love the idea of adding this extra loop but will add this for efficiency's sake

**Review by g-gaston — COMMENTED** [review]
/lgtm
